Feature request: rejecting not compatible keys if pinverification=1 is set

Is it possible to silently reject older keys unable to support pinverification=1 parameter?
Currently I am using several Yubikeys: NEO, 4 and 5 with two authentication methods:
with default u2f_keys file for authentication not requiring PIN covering all Yubikeys
with separate u2f_pin_keys file for PIN compatible keys for authentication where PIN is requried
In ideal scenario single u2f_keys file should be enough containing all the Yubikeys however if older key is inserted and pinverification=1 is specified authentication is denied.
Currently there appear first PIN prompt and only then authentication is rejected.

Hi,I have tried to build it but stumbled upon the following error during make:Making all in .make[1]: Entering directory '/home/percy/pam-u2f'&nbsp; CC&nbsp; &nbsp; &nbsp; &nbsp;pam-u2f.lo&nbsp; CC&nbsp; &nbsp; &nbsp; &nbsp;util.loutil.c: In function ‘do_authentication’:util.c:1340:43: warning: implicit declaration of function ‘fido_dev_supports_uv’; did you mean ‘fido_dev_supports_pin’? [-Wimplicit-function-declaration]&nbsp;1340 |&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;if (opts.uv == FIDO_OPT_FALSE &amp;&amp; !fido_dev_supports_uv(authlist[j]))&nbsp; &nbsp; &nbsp; |&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;^~~~~~~~~~~~~~~~~~~~&nbsp; &nbsp; &nbsp; |&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;fido_dev_supports_pinutil.c:1344:43: warning: implicit declaration of function ‘fido_dev_has_uv’; did you mean ‘fido_dev_has_pin’? [-Wimplicit-function-declaration]&nbsp;1344 |&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;(opts.uv == FIDO_OPT_TRUE &amp;&amp; !fido_dev_has_uv(authlist[j]))) {&nbsp; &nbsp; &nbsp; |&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;^~~~~~~~~~~~~~~&nbsp; &nbsp; &nbsp; |&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;fido_dev_has_pin&nbsp; CC&nbsp; &nbsp; &nbsp; &nbsp;drop_privs.lo&nbsp; CC&nbsp; &nbsp; &nbsp; &nbsp;b64.lo&nbsp; CC&nbsp; &nbsp; &nbsp; &nbsp;explicit_bzero.lo&nbsp; CCLD&nbsp; &nbsp; &nbsp;pam_u2f.la/bin/sh /home/percy/pam-u2f/build-aux/missing a2x --format=manpage -L -a revdate="Version 1.1.2" man/pam_u2f.8.txtmake[1]: Leaving directory '/home/percy/pam-u2f'Making all in pamu2fcfgmake[1]: Entering directory '/home/percy/pam-u2f/pamu2fcfg'gengetopt --no-handle-help --input cmdline.ggo Makefile.ammake&nbsp; all-ammake[2]: Entering directory '/home/percy/pam-u2f/pamu2fcfg'&nbsp; CC&nbsp; &nbsp; &nbsp; &nbsp;pamu2fcfg.o&nbsp; CC&nbsp; &nbsp; &nbsp; &nbsp;cmdline.o&nbsp; CC&nbsp; &nbsp; &nbsp; &nbsp;readpassphrase.o&nbsp; CC&nbsp; &nbsp; &nbsp; &nbsp;strlcpy.o&nbsp; CC&nbsp; &nbsp; &nbsp; &nbsp;../util.o../util.c: In function ‘do_authentication’:../util.c:1340:43: warning: implicit declaration of function ‘fido_dev_supports_uv’; did you mean ‘fido_dev_supports_pin’? [-Wimplicit-function-declaration]&nbsp;1340 |&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;if (opts.uv == FIDO_OPT_FALSE &amp;&amp; !fido_dev_supports_uv(authlist[j]))&nbsp; &nbsp; &nbsp; |&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;^~~~~~~~~~~~~~~~~~~~&nbsp; &nbsp; &nbsp; |&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;fido_dev_supports_pin../util.c:1344:43: warning: implicit declaration of function ‘fido_dev_has_uv’; did you mean ‘fido_dev_has_pin’? [-Wimplicit-function-declaration]&nbsp;1344 |&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;(opts.uv == FIDO_OPT_TRUE &amp;&amp; !fido_dev_has_uv(authlist[j]))) {&nbsp; &nbsp; &nbsp; |&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;^~~~~~~~~~~~~~~&nbsp; &nbsp; &nbsp; |&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;fido_dev_has_pin&nbsp; CC&nbsp; &nbsp; &nbsp; &nbsp;../b64.o&nbsp; CC&nbsp; &nbsp; &nbsp; &nbsp;../explicit_bzero.o&nbsp; CCLD&nbsp; &nbsp; &nbsp;pamu2fcfg/usr/bin/ld: ../util.o: in function `do_authentication':/home/percy/pam-u2f/pamu2fcfg/../util.c:1344: undefined reference to `fido_dev_has_uv'/usr/bin/ld: /home/percy/pam-u2f/pamu2fcfg/../util.c:1340: undefined reference to `fido_dev_supports_uv'collect2: error: ld returned 1 exit statusmake[2]: *** [Makefile:455: pamu2fcfg] Error 1make[2]: Leaving directory '/home/percy/pam-u2f/pamu2fcfg'make[1]: *** [Makefile:360: all] Error 2make[1]: Leaving directory '/home/percy/pam-u2f/pamu2fcfg'make: *** [Makefile:794: all-recursive] Error 1Temat: Re: [Yubico/pam-u2f] Feature request: rejecting not compatible keys if pinverification=1 is set (#183)Data: 2021-08-30 15:38Nadawca: "Ludvig Michaelsson" Adresat: "Yubico/pam-u2f" ; DW: "perceival" ; "Mention" ; 
Hi @perceival,
I have pushed a tentative implementation of your feature request to the opts branch of this repository. If you could build it from the source and temporarily point PAM to that build of pam-u2f to see if it satisfies your use-case, that would be greatly appreciated.
Thank you!

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android.

--- a/util.c
+++ b/util.c
@@ -1029,6 +1029,57 @@ static void parse_opts(const cfg_t *cfg, const char *attr, struct opts *opts) {
   }
 }
 
+static int get_device_opts(fido_dev_t *dev, int *pin, int *uv) {
+  fido_cbor_info_t *info = NULL;
+  char *const *ptr;
+  const bool *val;
+  size_t len;
+
+  *pin = *uv = -1; /* unsupported */
+
+  if (fido_dev_is_fido2(dev)) {
+    if ((info = fido_cbor_info_new()) == NULL ||
+        fido_dev_get_cbor_info(dev, info) != FIDO_OK) {
+      fido_cbor_info_free(&info);
+      return 0;
+    }
+
+    ptr = fido_cbor_info_options_name_ptr(info);
+    val = fido_cbor_info_options_value_ptr(info);
+    len = fido_cbor_info_options_len(info);
+    for (size_t i = 0; i < len; i++) {
+      if (strcmp(ptr[i], "clientPin") == 0) {
+        *pin = val[i];
+      } else if (strcmp(ptr[i], "uv") == 0) {
+        *uv = val[i];
+      }
+    }
+  }
+
+  fido_cbor_info_free(&info);
+  return 1;
+}
+
+static int match_device_opts(fido_dev_t *dev, struct opts *opts) {
+  int pin, uv;
+
+  /* FIXME: fido_dev_{supports,has}_{pin,uv} (1.7.0) */
+  if (!get_device_opts(dev, &pin, &uv)) {
+    return -1;
+  }
+
+  if (opts->uv == FIDO_OPT_FALSE && uv < 0) {
+    opts->uv = FIDO_OPT_OMIT;
+  }
+
+  if ((opts->pin == FIDO_OPT_TRUE && pin != 1) ||
+      (opts->uv == FIDO_OPT_TRUE && uv != 1)) {
+    return 0;
+  }
+
+  return 1;
+}
+
 static int set_opts(const cfg_t *cfg, const struct opts *opts,
                     fido_assert_t *assert) {
   if (fido_assert_set_up(assert, opts->up) != FIDO_OK) {
@@ -1331,12 +1382,22 @@ int do_authentication(const cfg_t *cfg, const device_t *devices,
       goto out;
     }
 
-    /* options used during authentication */
-    parse_opts(cfg, devices[i].attributes, &opts);
-
     if (get_authenticators(cfg, devlist, ndevs, assert,
                            is_resident(devices[i].keyHandle), authlist)) {
       for (size_t j = 0; authlist[j] != NULL; j++) {
+        /* options used during authentication */
+        parse_opts(cfg, devices[i].attributes, &opts);
+
+        r = match_device_opts(authlist[j], &opts);
+        if (r != 1) {
+          if (cfg->debug) {
+            D(cfg->debug_file, "%s, skipping authenticator",
+              r < 0 ? "Failed to query supported options"
+                    : "Unsupported options");
+          }
+          continue;
+        }
+
         if (!set_opts(cfg, &opts, assert)) {
           if (cfg->debug)
             D(cfg->debug_file, "Failed to set assert options");

